File: Prolog/diff.pl
% diff( +Expr, +Atom, -Expr ).

diff( X, X, 1 ).

diff( C, _X, 0 ) :-
        number(C).

diff( Y, X, 0 ) :- 
        atom(Y),
        X \= Y.

diff( - F, X, - DF ) :-
        diff( F, X, DF ).

diff( F + G, X, DF + DG ) :- 
        diff( F, X, DF ), 
        diff( G, X, DG ).

diff( F - G, X, DF - DG) :- 
        diff( F, X, DF ), 
        diff( G, X, DG ).

diff( F * G, X, DF * G + F * DG ) :- 
        diff( F, X, DF ), 
        diff( G, X, DG ).

diff( F / G, X, (DF * G - F * DG) / (G * G) ) :- 
        diff( F, X, DF ),
        diff( G, X, DG ).

diff( F ** N, X, N * DF * F ** N1 ) :-
        number(N),
        N1 is N - 1,
        diff( F, X, DF ).

diff( F ** G, X, D ) :-
        diff( exp(ln(F) * G), X, D ).

diff( exp(F), X, DF * exp(F) ) :- 
        diff( F, X, DF ).

diff( ln(F), X, DF / F ) :- 
        diff( F, X, DF ).

File: Python/diff.py
import exprParser as ep

def diff(e):
    "differentiate the expressions e with respect to the variable x"
    if e[0] == '+':
        f , g  = e[1:]
        fs, gs = diff(f), diff(g)
        return ('+', fs, gs)
    if e[0] == '-':
        f , g  = e[1:]
        fs, gs = diff(f), diff(g)
        return ('-', fs, gs)
    if e[0] == '*':
        f , g  = e[1:]
        fs, gs = diff(f), diff(g)
        return ('+', ('*', fs, g), ('*', f, gs))
    if e[0] == '/':
        f , g  = e[1:]
        fs, gs = diff(f), diff(g)
        return ('/', ('-', ('*', fs, g), ('*', f, gs)), ('*', g, g))
    if e[0] == '**':
        f , g  = e[1:]
        return diff(('exp', ('*', g, ('ln', f))))
    if e[0] == 'ln':
        f  = e[1]
        fs = diff(f) 
        return ('/', fs, f)
    if e[0] == 'exp':
        f  = e[1]
        fs = diff(f) 
        return ('*', fs, e)
    if e[0] == 'x':
        return '1'
    return 0

def test(s):
    t = ep.ExprParser(s).parse()
    d = diff(t)
    print("d/dx", s, "=", ep.toString(d))

if __name__ == "__main__":
    test("x")
    test("y")
    test("1")
    test("x+x")
    test("x+x+y")
    test("x*x")
    test("x/x")
    test("1/x")
    test("ln(x)")
    test("exp(x)")
    test("x ** x")
    
